clippy_lints/src/unwrap.rs

```diff
@@ -0,0 +1,176 @@
+use rustc::lint::*;
+
+use rustc::hir::intravisit::*;
+use rustc::hir::*;
+use syntax::ast::NodeId;
+use syntax::codemap::Span;
+use utils::{in_macro, match_type, paths, usage::is_potentially_mutated};
+
+/// **What it does:** Checks for calls of unwrap[_err]() that cannot fail.
+///
+/// **Why is this bad?** Using `if let` or `match` is more idiomatic.
+///
+/// **Known problems:** Limitations of the borrow checker might make unwrap() necessary sometimes?
+///
+/// **Example:**
+/// ```rust
+/// if option.is_some() {
+///     do_something_with(option.unwrap())
+/// }
+/// ```
+///
+/// Could be written:
+///
+/// ```rust
+/// if let Some(value) = option {
+///     do_something_with(value)
+/// }
+/// ```
+declare_clippy_lint! {
+    pub UNNECESSARY_UNWRAP,
+    complexity,
```

Comment:
let's put it in the `nursery` for now until we have some data on larger projects